rtape.c

```diff
@@ -0,0 +1,647 @@
+#define FLG_HER    000200  /* Hard error */
+#define FLG_SER    000400  /* Soft error */
+#define FLG_OFFL   001000  /* Offline */
+#define FLG_NREC   002000  /* Non-record-oriented device */
+/* Not part of status response. */
+#define FLG_NOREW  00200000
+#define FLG_WRITE  00400000
+
+#define MIN(X, Y)  ((X) < (Y) ? (X) : (Y))
+
+static unsigned char command_data[MAX_RECORD + 3];
+static int command_opcode;
+static int command_len;
+static unsigned char *command_ptr;
+static int flags;
+static int allow_slash = 0;
+static int daemonize = 0;
+static int read_only = 0;
+
+#define VERSION 1
+static const char *contact = "RTAPE";
+static const char *version = "RECORD STREAM VERSION 1\215";
+
+static void serve(void);
+static void send_packet(int opcode, const void *data, size_t len);
+static void send_status(int id, const char *message);
+
+typedef void handler_t(const unsigned char *data, int len);
+
+static handler_t packet_rfc;
+static handler_t packet_los;
+static handler_t packet_cls;
+static handler_t packet_dat;
+
+static handler_t state_ignore;
+static handler_t state_version;
+static handler_t state_opcode;
+static handler_t state_len1;
+static handler_t state_len2;
+static handler_t state_data;
+static handler_t *state;
+
+static handler_t cmd_login;
+static handler_t cmd_mount;
+static handler_t cmd_probe;
+static handler_t cmd_read;
+static handler_t cmd_write;
+static handler_t cmd_space_file;
+static handler_t cmd_space_record;
+static handler_t cmd_rewind;
+static handler_t cmd_write_mark;
+static handler_t cmd_close;
+
+struct handler {
+  int opcode;
+  handler_t *fn;
+};
+
+struct handler packet_handler[] = {
+  { CHOP_RFC, packet_rfc },
+  { CHOP_LOS, packet_los },
+  { CHOP_CLS, packet_cls },
+  { CHOP_DAT, packet_dat }
+};
+
+#define MAX_HANDLERS (sizeof packet_handler / sizeof packet_handler[0])
+
+struct handler command_handler[] = {
+  { CMD_LGI, cmd_login },
+  { CMD_MNT, cmd_mount },
+  { CMD_PRB, cmd_probe },
+  { CMD_RD,  cmd_read },
+  { CMD_WRT, cmd_write },
+  { CMD_RWD, cmd_rewind },
+  { CMD_SPF, cmd_space_file },
+  { CMD_SPR, cmd_space_record },
+  { CMD_WFM, cmd_write_mark },
+  { CMD_CLS, cmd_close }
+};
+
+#define MAX_COMMANDS (sizeof command_handler / sizeof command_handler[0])
+
+static char chaos_socket_directory[] = "/tmp";
+static FILE *log;
+static int sock;
+static int tape;
+
+static void dispatch(int opcode, int n, struct handler *handler,
+                     const unsigned char *data, int len)
+{
+  int i;
+  for (i = 0; i < n; i++) {
+    if (opcode == handler[i].opcode) {
+      handler[i].fn(data, len);
+      return;
+    }
+  }
+  fprintf(stderr, "Unknown operation: %d\n", opcode);
+  exit(1);
+}
+
+static int
+connect_to_named_socket(int socktype, char *path)
+{
+  int sock, slen;
+  struct sockaddr_un local, server;
+  
+  local.sun_family = AF_UNIX;
+  sprintf(local.sun_path, "%s/%s_%d", chaos_socket_directory, path, getpid());
```

Comment:
I suggest skipping the binding of "this end" of the socket, since is isn't necessary and just leaves a lot of garbage sockets in /tmp. (I had such code in my early versions, until I realised that.)